### PR TITLE
Nokogiri patches

### PR DESCRIPTION
**Pull request by aardvark179**
In combination with the `nokogiri-memory-management` fixes allow most nokogiri tests to pass and other gems requiring it to be tested.

### CODE REVIEW

File: lib/cext/preprocess.rb [hunk 1]
```diff
@@ -95,8 +95,168 @@
...
+                               }
+NOKOGIRI_WRAP_ERROR_PATCH_1 = { match: 'VALUE msg',
+                                replacement: 'VALUE msg[1]'
+                              }
+NOKOGIRI_WRAP_ERROR_PATCH_2 = { match: 'msg = ',
+                                replacement: 'msg[0] = '
+                              }
+NOKOGIRI_WRAP_ERROR_PATCH_3 = { match: '&msg',
+                                replacement: 'msg'
+                              }
```
Comment (eregon) [inline lib/cext/preprocess.rb]: If these constants are only used in one place below, could you inline them?
It makes reading them significantly harder.

Comment (eregon) [reply lib/cext/preprocess.rb]: The `NO_ASSIGNMENT` constant should remain last, next to `read_field`.

File: lib/cext/preprocess.rb [hunk 2]
```diff
@@ -95,8 +95,168 @@
...
+                              replacement: "rb_tr_managed_from_handle_or_null(xsltGetExtData(transform, functionURI));"
+                            }
+
+NOKOGIRI_NODE_PRIVATE_CAST_PATCH = { match: "(VALUE)node->_private;",
+                              replacement: "rb_tr_managed_from_handle_or_null(node->_private);"
+                            }
+
+NOKOGIRI_CTX_STYLE_CAST_PATCH = { match: "ctxt->style->_private",
+                              replacement: "rb_tr_managed_from_handle_or_null(ctxt->style->_private)"
+                            }
```
Comment (eregon) [inline lib/cext/preprocess.rb]: Can you use the `read_write_field` helper for such cases?

**Comment by eregon** [comment]
@aardvark179 Could you inine the rest of the constants, unless they are multi-line strings?
It's no fun to scroll and match manually.

File: lib/cext/preprocess.rb [hunk 3]
```diff
@@ -93,10 +93,32 @@
...
+static int dealloc_node_i(xmlNodePtr key, xmlNodePtr node, xmlDocPtr doc)
+{
+EOF
+
+NOKOGIRI_DEALLOC_DECL_NEW = <<-EOF
+static int dealloc_node_i(st_data_t a, st_data_t b, st_data_t c, int errorState)
+{
+  xmlNodePtr key = (xmlNodePtr)a;
+  xmlNodePtr node = (xmlNodePtr)b;
+  xmlDocPtr doc = (xmlDocPtr)c;
```
Comment (eregon) [inline lib/cext/preprocess.rb]: Is this to avoid warnings turning into errors?
Maybe we can adjust the flags, like in openssl's extconf.rb, but patching might still be best to catch potential problems early.

File: lib/cext/preprocess.rb [hunk 4]
```diff
@@ -131,20 +153,268 @@ def write_array(name, leaking)
...
+      cast_native_for_value('node->_private', ';'),
+      cast_value_for_native('rb_node'),
+   ]
+  },
+  'xml_namespace.c' => {
+    gem: 'nokogiri',
+    patches: [
+      { match: NOKOGIRI_DOC_RUBY_OBJECT_ORIG,
+                                  replacement: NOKOGIRI_DOC_RUBY_OBJECT_NEW
+                                },
```
Comment (eregon) [inline lib/cext/preprocess.rb]: alignment

File: lib/cext/preprocess.rb [hunk 5]
```diff
@@ -131,20 +153,268 @@ def write_array(name, leaking)
...
+  'xml_sax_parser.c' => {
+    gem: 'nokogiri',
+    patches: [
+      {
+        match: 'NOKOGIRI_SAX_SELF(ctx)',
+        replacement: 'rb_tr_managed_from_handle_or_null(NOKOGIRI_SAX_SELF(ctx))'
+      },
+      {
+        match: /va_list args;[^}]*id_warning, 1, ruby_message\);/,
+        replacement: 'rb_funcall(doc, id_warning, 1, NOKOGIRI_STR_NEW2("Warning."));'
```
Comment (eregon) [inline lib/cext/preprocess.rb]: Why is this needed? `vasprintf` doesn't work? Can we fix it?

File: lib/cext/preprocess.rb [hunk 6]
```diff
@@ -131,20 +153,268 @@ def write_array(name, leaking)
 end
 
 def read_write_array(name, leaking)
-  [read_array(name), 
+  [read_array(name),
    write_array(name, leaking)]
 end
 
+def cast_value_for_native(name, suffix = '', type = :local)
```
Comment (eregon) [inline lib/cext/preprocess.rb]: What does this function solve? Could you add a comment (same for other functions)?

File: lib/cext/preprocess.rb [hunk 7]
```diff
@@ -131,20 +153,268 @@ def write_array(name, leaking)
...
+        replacement: 'VALUE msg[1]'
+      },
+      {
+        match: 'msg = ',
+        replacement: 'msg[0] = '
+      },
+      {
+        match: '&msg',
+        replacement: 'msg'
+      },
```
Comment (eregon) [inline lib/cext/preprocess.rb]: Would force inlining `rb_class_new_instance` be enough for that case?

File: lib/cext/preprocess.rb [hunk 8]
```diff
@@ -131,20 +153,268 @@ def write_array(name, leaking)
...
+        replacement: "VALUE thing = Qnil;\n  VALUE errors = rb_ary_new();"
+      },
+      {
+        match: 'xmlSetStructuredErrorFunc(NULL, Nokogiri_error_raise);',
+        replacement: 'xmlSetStructuredErrorFunc(rb_tr_handle_for_managed(errors), Nokogiri_error_array_pusher);'
+      },
+      {
+        match: 'if(xpath == NULL)',
+        replacement: "if (RARRAY_LEN(errors) > 0) { rb_exc_raise(rb_ary_entry(errors, 0)); }\nif(xpath == NULL)"
+      },
```
Comment (eregon) [inline lib/cext/preprocess.rb]: Do we need a Ruby array or a simple variable or a C array would be enough here?

File: lib/cext/preprocess.rb [hunk 9]
```diff
@@ -131,20 +153,268 @@ def write_array(name, leaking)
...
+      cast_native_for_value('(ctx->context->userData)', ';'),
+      { match: 'VALUE *argv', replacement: 'VALUE argv[32]' },
+      { match: /argv =.*$/, replacement: '' },
```
Comment (eregon) [inline lib/cext/preprocess.rb]: These two sound like bad hacks :p

File: lib/cext/preprocess.rb [hunk 10]
```diff
@@ -131,20 +153,268 @@ def write_array(name, leaking)
...
+      {
+        match: 'xmlSetStructuredErrorFunc(NULL, Nokogiri_error_raise);',
+        replacement: 'xmlSetStructuredErrorFunc(rb_tr_handle_for_managed(errors), Nokogiri_error_array_pusher);'
+      },
+      {
+        match: 'if(xpath == NULL)',
+        replacement: "if (RARRAY_LEN(errors) > 0) { rb_exc_raise(rb_ary_entry(errors, 0)); }\nif(xpath == NULL)"
+      },
+      cast_native_for_value('(ctx->context->userData)', ';'),
+      { match: 'VALUE *argv', replacement: 'VALUE argv[32]' },
```
Comment (eregon) [inline lib/cext/preprocess.rb]: This would also match
https://github.com/sparklemotion/nokogiri/blob/v1.8.1/ext/nokogiri/xml_xpath_context.c#L190
intended? Probably you want an extra `;`.

File: lib/cext/preprocess.rb (hunk 3 above)
Comment (aardvark179) [reply lib/cext/preprocess.rb]: Yes, they play a little fast and loose with the function signature and when I was producing the patch set I wanted to be careful we didn't trip up on things like that.

File: lib/cext/preprocess.rb [hunk 11]
```diff
@@ -93,10 +93,32 @@
...
+
+NOKOGIRI_DEALLOC_DECL_NEW = <<-EOF
+static int dealloc_node_i(st_data_t a, st_data_t b, st_data_t c, int errorState)
+{
+  xmlNodePtr key = (xmlNodePtr)a;
+  xmlNodePtr node = (xmlNodePtr)b;
+  xmlDocPtr doc = (xmlDocPtr)c;
+EOF
+
+ID = /([a-zA-Z0-9_]+)/
```
[the remark this answers is not in the document]
Comment (aardvark179) [reply lib/cext/preprocess.rb]: You're right, and it shouldn't be quite the same as `\w` because the first character cannot be a number. I'll fix that.

File: lib/cext/preprocess.rb [hunk 12]
```diff
@@ -131,20 +153,268 @@ def write_array(name, leaking)
...
   'xml_node_set.c' => {
     gem: 'nokogiri',
     patches: [
       {
         match: /[[:blank:]]*?switch\s*?\(.*?Qnil:/m,
         replacement: XML_NODE_SET_PATCH
+      },
+      {
+        match: 'static VALUE to_array(VALUE self, VALUE rb_node)',
+        replacement: 'static VALUE to_array(VALUE self)'
```
[the remark this answers is not in the document]
Comment (aardvark179) [reply lib/cext/preprocess.rb]: Yes. Turns out the MRI and most C ABIs won't complain if you pass too few arguments to a function, and everything works as long as you never use them. We are more careful.

File: lib/cext/preprocess.rb (hunk 5 above)
Comment (aardvark179) [reply lib/cext/preprocess.rb]: Var args don't work for callbacks from native code to Sulong. I discussed this a bit with Roland and we couldn't see a good way to make it work without some major work to both Sulong and our libffi code.

File: lib/cext/preprocess.rb (hunk 8 above)
Comment (aardvark179) [reply lib/cext/preprocess.rb]: There's already an error handling function for pushing the errors to a ruby array, which is why I used it.

File: lib/cext/preprocess.rb (hunk 10 above)
Comment (aardvark179) [reply lib/cext/preprocess.rb]: Good catch. I'm amazed that compiled and worked really. :grin:

File: lib/cext/preprocess.rb (hunk 9 above)
Comment (aardvark179) [reply lib/cext/preprocess.rb]: I agree they are. When I patched nokogiri by hand I turned that whole thing into a ruby array, but it's a much more invasive patch for that function.. I've now got that more complex version working so am including it.

File: lib/cext/preprocess.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (eregon) [reply lib/cext/preprocess.rb]: Please add a comment next to the `{` above the `match:` explaining this and other non-trivial patches.

File: lib/cext/preprocess.rb (hunk 5 above)
Comment (eregon) [reply lib/cext/preprocess.rb]: In that case, could we use the native vasprintf and force every argument to go native?
